ironfish/src/memPool/recentlyEvictedCache.ts

```diff
@@ -0,0 +1,225 @@
+
+  private currentSequence = 0
+
+  /**
+   * A priority queue of transactions to evict from this cache when full.
+   *
+   * Transactions are evicted in order of `feeRate` descending because
+   * they are most likely to be re-included into the mempool
+   */
+  private readonly evictionQueue: PriorityQueue<EvictionEntry>
+
+  private readonly expirationQueue: PriorityQueue<ExpirationEntry>
+
+  /**
+   * Creates a new cache to track transactions that have recently been evicted.
+   * Transactions are routinely evicted from the cache after the `maxJailTime`.
+   * When full, transactions are evicted in order of `feeRate` descending to
+   * make room for the new transaction.
+   * @param options
+   */
+  constructor(options: { metrics: MetricsMonitor; logger?: Logger }) {
+    const logger = options.logger || createRootLogger()
+
+    this.metrics = options.metrics
+    this.logger = logger.withTag('RecentlyEvictedCache')
+
+    this.evictionQueue = new PriorityQueue<EvictionEntry>(
+      (t1, t2) => t1.feeRate > t2.feeRate,
+      (t) => t.hash.toString('hex'),
+    )
+
+    this.expirationQueue = new PriorityQueue<ExpirationEntry>(
+      (t1, t2) => t1.expiration < t2.expiration,
+      (t) => t.hash.toString('hex'),
+    )
+  }
+
+  private remove(hash: TransactionHash): Transaction | void {
+    const transactionToRemove = this.get(hash)
+
+    if (!transactionToRemove) {
+      return
+    }
+
+    this.transactions.delete(hash)
+
+    this.transactionsBytes -= getTransactionSize(transactionToRemove)
+
+    return transactionToRemove
+  }
+
+  /**
+   * Removes the transaction with the highest fee rate from the cache.
+   *
+   * If the transaction cache and eviction queue are out of sync, this method
+   * lazily pops from the eviction queue until it finds a transaction that is
+   * present in the cache
+   *
+   * @returns the transaction that was evicted, or undefined if the cache is empty
+   */
+  private poll(): Transaction | void {
+    let hashToRemove: Buffer | undefined
+
+    while (!this.isEmpty() && this.evictionQueue.size() > 0) {
+      const toEvict = this.evictionQueue.poll()
+
+      Assert.isNotUndefined(toEvict)
+
+      // lazy deletion
+      if (!this.has(toEvict.hash)) {
+        continue
+      }
+
+      hashToRemove = toEvict.hash
+      break
+    }
+
+    if (!hashToRemove) {
+      return
+    }
+
+    return this.remove(hashToRemove)
+  }
+
+  count(): number {
+    return this.transactions.size
+  }
+
+  isFull(): boolean {
+    return this.count() === this.capacity
+  }
+
+  isEmpty(): boolean {
+    return this.count() === 0
+  }
+
+  /**
+   * Adds a new transaction to the recently evicted cache.
+   * If the cache is full, the transaction with the highest fee rate will be evicted.
+   *
+   */
+  add(transaction: Transaction): void {
+    const hash = transaction.hash()
+
+    if (this.has(hash)) {
+      return
+    }
+
+    if (this.isFull()) {
```

Comment:
Do we want to add first then evict? So the current transaction `feeRate` is taken into account